internal/graph/check_fast_path.go

```diff
@@ -994,10 +985,119 @@ func (c *LocalChecker) recursiveTTUFastPath(ctx context.Context, req *ResolveChe
 	ctx, span := tracer.Start(ctx, "recursiveTTUFastPath")
 	defer span.End()
 
+	leftChannelBuilder := &simpleRecursiveObjectProvider{}
+
 	return c.recursiveFastPath(ctx, req, iter, &recursiveMapping{
 		kind:             storage.TTUKind,
 		tuplesetRelation: rewrite.GetTupleToUserset().GetTupleset().GetRelation(),
-	})
+	}, leftChannelBuilder)
+}
+
+// objectProvider is an interface that abstracts the building of a channel that holds object IDs or usersets.
+type objectProvider interface {
+	close()
+	build(ctx context.Context, req *ResolveCheckRequest) (chan usersetMessage, error)
+}
+
+type simpleRecursiveObjectProvider struct {
+	mapper storage.TupleMapper
+}
+
+var _ objectProvider = &simpleRecursiveObjectProvider{}
+
+func (s *simpleRecursiveObjectProvider) close() {
+	if s.mapper != nil {
+		s.mapper.Stop()
+	}
+}
+
+func (s *simpleRecursiveObjectProvider) build(ctx context.Context, req *ResolveCheckRequest) (chan usersetMessage, error) {
```

Comment:
FYI the code in this function was previously within `recursiveFastPath`

Reply:
It's half new half old. The majority of the implementation existed, its now encapsulated differently.

Reply:
Do we need to have a `simple` and `complex` implementation? Seems to me the "complex" one can resolve the simple one, ultimately it's the same cost/underlying implementation.

Reply:
> It's half new half old. The majority of the implementation existed, its now encapsulated differently.

WDYM, "half new half old"? I didn't change how it works, I just put it somewhere else.

Reply:
>  I wrote it to keep v1 code working.

We not keep v1 completely untouched?

Reply:
Yes, but note that if the experimental is enabled, all code that used to go thru v1 now will go through v2 (which is a generalization of v1)